File: ui/src/api/procurementLabels.ts
import { api } from "./client";

export interface ProcurementLabelRow {
  line_no: number;
  onec_item_code: string;
  item_name: string;
  article_1c: string;
  sku: string;
  barcode: string;
  barcode_source: string;
  unit: string;
  quantity: string;
  price?: string | null;
  amount?: string | null;
  certificate_id: string;
  certificate_item_id: string;
  certificate_number: string;
  certificate_valid_to: string;
  certificate_file_id: string;
  certificate_status: string;
  eac_allowed: boolean;
  product_passport_item_id: string;
  trade_name: string;
  tnved: string;
  manufacturer: string;
  product_series: string;
  label_warnings: string[];
  status: string;
  blockers: string[];
}

export interface ProcurementLabelPreview {
  item_id: string;
  entity_type_id: number;
  onec_number: string;
  title: string;
  contour: string;
  status: string;
  ready: boolean;
  blocked: boolean;
  blockers: string[];
  rows: ProcurementLabelRow[];
  artifact_version?: number | null;
  zip_url?: string | null;
  disk_file_id?: string | null;
}

export interface ProcurementLabelGenerateResponse {
  preview: ProcurementLabelPreview;
  generated: boolean;
  artifact_version?: number | null;
  zip_filename?: string | null;
  zip_url?: string | null;
  disk_file_id?: string | null;
}

export interface ProcurementCertificationDocsGenerateResponse extends ProcurementLabelGenerateResponse {
  gtin_rows: number;
  missing_rows: number;
  document_checklist: string[];
}

export async function fetchProcurementLabelPreview(itemId: string) {
  const { data } = await api.get<ProcurementLabelPreview>(
    `/procurement-labels/orders/${encodeURIComponent(itemId)}/preview`
  );
  return data;
}

export async function generateProcurementLabels(itemId: string) {
  const { data } = await api.post<ProcurementLabelGenerateResponse>(
    `/procurement-labels/orders/${encodeURIComponent(itemId)}/generate`,
    { dry_run: false }
  );
  return data;
}

export async function generateProcurementCertificationDocs(itemId: string) {
  const { data } = await api.post<ProcurementCertificationDocsGenerateResponse>(
    `/procurement-labels/orders/${encodeURIComponent(itemId)}/certification-docs/generate`,
    { dry_run: false }
  );
  return data;
}

export async function approveProcurementLabels(itemId: string) {
  const { data } = await api.post<{
    item_id: string;
    status: string;
    artifact_version?: number | null;
    zip_url?: string | null;
  }>(`/procurement-labels/orders/${encodeURIComponent(itemId)}/approve`, {});
  return data;
}

export async function sendProcurementLabelsToFactory(itemId: string) {
  const { data } = await api.post<{
    item_id: string;
    status: string;
    artifact_version?: number | null;
    zip_url?: string | null;
  }>(`/procurement-labels/orders/${encodeURIComponent(itemId)}/send-to-factory`, {});
  return data;
}

File: app/schemas/procurement_labels.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field


class ProcurementLabelsSessionRequest(BaseModel):
    access_token: str = Field(min_length=1)
    domain: str = Field(min_length=1)
    member_id: str = Field(min_length=1)


class ProcurementLabelsUser(BaseModel):
    user_id: str
    name: str | None = None


class ProcurementLabelsSessionResponse(BaseModel):
    session_token: str
    token_type: str = "bearer"
    expires_at: datetime
    expires_in: int
    user: ProcurementLabelsUser


class ProcurementLabelRow(BaseModel):
    line_no: int
    onec_item_code: str
    item_name: str
    article_1c: str = ""
    sku: str
    barcode: str
    barcode_source: str = ""
    unit: str
    quantity: Decimal
    price: Decimal | None = None
    amount: Decimal | None = None
    certificate_id: str = ""
    certificate_item_id: str = ""
    certificate_number: str = ""
    certificate_valid_to: str = ""
    certificate_file_id: str = ""
    certificate_status: str
    eac_allowed: bool
    product_passport_item_id: str = ""
    trade_name: str = ""
    tnved: str = ""
    manufacturer: str = ""
    product_series: str = ""
    label_warnings: list[str] = Field(default_factory=list)
    status: str
    blockers: list[str] = Field(default_factory=list)


class ProcurementLabelOrderPreview(BaseModel):
    item_id: str
    entity_type_id: int
    onec_number: str
    title: str
    contour: str
    status: str
    ready: bool
    blocked: bool
    blockers: list[str] = Field(default_factory=list)
    rows: list[ProcurementLabelRow] = Field(default_factory=list)
    artifact_version: int | None = None
    zip_url: str | None = None
    disk_file_id: str | None = None


class ProcurementLabelGenerateRequest(BaseModel):
    dry_run: bool = False


class ProcurementLabelGenerateResponse(BaseModel):
    preview: ProcurementLabelOrderPreview
    generated: bool
    artifact_version: int | None = None
    zip_filename: str | None = None
    zip_url: str | None = None
    disk_file_id: str | None = None


class ProcurementCertificationDocsGenerateResponse(BaseModel):
    preview: ProcurementLabelOrderPreview
    generated: bool
    artifact_version: int | None = None
    zip_filename: str | None = None
    zip_url: str | None = None
    disk_file_id: str | None = None
    gtin_rows: int = 0
    missing_rows: int = 0
    document_checklist: list[str] = Field(default_factory=list)


class ProcurementLabelApproveResponse(BaseModel):
    item_id: str
    status: str
    artifact_version: int | None = None
    zip_url: str | None = None
